frostsnap_core/src/device_nonces.rs

```diff
@@ -0,0 +1,379 @@
+use alloc::collections::VecDeque;
+use alloc::vec::Vec;
+use chacha20::{
+    cipher::{KeyIvInit, StreamCipher},
+    ChaCha20,
+};
+use rand_core::RngCore;
+use schnorr_fun::{
+    binonce,
+    frost::{NonceKeyPair, PairedSecretShare, PartySignSession, SignatureShare},
+    fun::prelude::*,
+};
+
+use crate::{
+    nonce_stream::{CoordNonceStreamState, NonceStreamId, NonceStreamSegment},
+    SignSessionId, NONCE_BATCH_SIZE,
+};
+
+type ChaChaSeed = [u8; 32];
+
+#[derive(Clone, Debug, PartialEq, bincode::Encode, bincode::Decode)]
+pub struct SecretNonceSlot {
+    //XXX: index must be first so we can decode it easily without decoding the whole thing
+    //XXX: index == u32::MAX is invalid because we use it to represent "uninitialized"
+    pub index: u32,
+    pub nonce_stream_id: NonceStreamId,
+    pub ratchet_prg_seed: ChaChaSeed,
+    pub signing_state: Option<SigningState>,
+}
+
+#[derive(Clone, Debug, PartialEq, bincode::Encode, bincode::Decode)]
+pub struct SigningState {
+    pub session_id: SignSessionId,
+    pub signature_shares: Vec<SignatureShare>,
+}
+
+pub trait NonceStreamSlot {
+    fn read_index(&mut self) -> Option<u32>;
+    fn read_slot(&mut self) -> Option<SecretNonceSlot>;
+    fn write_slot(&mut self, value: &SecretNonceSlot);
+}
+
+/// Manages two writable sectors of persistent storage such that we make sure the state of the system we're managing is never lost.
+/// The new state is first written, if that succeeds we finally write over the previous state.
+/// In this case we're managing the secret seed needed to produce nonces for signatures.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ABSlot<S> {
+    slots: [S; 2],
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ABSlots<S> {
+    slots: Vec<ABSlot<S>>,
+}
+
+impl<S: NonceStreamSlot> ABSlots<S> {
+    pub fn new(ab_slots: impl IntoIterator<Item = ABSlot<S>>) -> Self {
+        Self {
+            slots: ab_slots.into_iter().collect(),
+        }
+    }
+
+    pub fn get_or_create(
+        &mut self,
+        stream_id: NonceStreamId,
+        rng: &mut impl RngCore,
+    ) -> &mut ABSlot<S> {
+        let mut i = 0;
+        let found = loop {
+            let ab_slot = &mut self.slots[i];
+            match ab_slot.nonce_stream_id() {
+                Some(slot_nsid) => {
+                    if slot_nsid == stream_id {
+                        break Some(i);
+                    }
+                }
+                None => {
+                    ab_slot.initialize(stream_id, rng);
```

Comment:
Might be worth noting that this design allocates slots sequentially, we never delete them and overwrite them.
This means that searching & initializing like this is OK since we'll never have an empty slot before a non-empty one(s).

Reply:
Will add a note about this in the subsequent PR since this has #208 building on it